docs/src/design/proposals/Registration.md

```diff
@@ -0,0 +1,518 @@
+        .notify("Hello world!");
+    });
+
+    // Code in the constructor is run when the plugin is instantiated;
+    // any extension points exposed by the plugin should be declared
+    // here, typically as providers.
+    this.someRegistry = new mct.Registry();
+  }
+
+  MyPlugin.prototype = Object.create(Plugin.prototype);
+
+  return MyPlugin;
+
+});
+```
+
+## Extending Extensions
+
+Augmenting the behavior of individual extensions is difficult in the
+existing bundle mechanism; for instance, to decorate a specific
+Action one has to decorate the ActionService, recognize the specific
+Action, and then decorate _that_.
+
+By introducing `Provider`s as a more general pattern for extensibility,
+cases like this may be simplified, provided that an appropriate `Plugin`
+exposes individual extensions in a suitable way.
+
+As an example, a plugin implementing a "navigate" action might look like:
+
+```
+function NavigationPlugin(core) {
+    var navigateActionProvider = new Provider(function () {
+        return new NavigateAction();
+    });
+
+    // Expose to permit extension
+    this.navigateActionProvider = navigateActionProvider;
+
+    Plugin.call(this, function () {
+        core.actionRegistry.register(function () {
+            return navigateActionProvider.get();
+        });
+    });
+}
+```
+
+And other plugins could then extend the "navigate" action directly:
+
+```
+function EditPlugin(navigationPlugin) {
+    Plugin.call(this, function () {
+        navigationPlugin.navigateActionProvider
+            .decorate(function (navigateAction) {
+                return new DecoratedNavigationAction(navigateAction);
+            });
+    });
+}
+```
+
+# Legacy Support
+
+The proposed API does not achieve parity with the existing Framework
+Layer. In particular:
+
+1. There is no Angular integration.
+2. Transitional support for legacy bundles is not specified.
+3. Script loading is not handled.
+
+Propose to resolve these by:
+
+1. Adding an Angular integration plugin that uses registries for
+   controllers/directives/etc. and maps these to appropriate configuration
+   of an Angular application.
+2. Adding a legacy support plugin that maps (and transforms, as necessary)
+   extensions exposed by legacy bundles to appropriate registries and/or
+   service providers using the new Registration API.
+3. Not handling script loading; the Registration API expects objects,
+   so how the scripts necessary to instantiate those objects (and how they
+   are instantiated) is left to the application developer.
+
+# Evaluation
+
+[Identified problems](#problems-to-address) are addressed by this solution:
+
+1. Dependencies between plugins can be made explicit; a `Plugin` may
+   impose dependencies on other specific `Plugin` subclasses as constructor
+   arguments, disambiguated with JSDoc. The software does not take part
+   in dependency management among plugins; rather, this responsibility is
+   plainly communicated to developers.
```

Comment:
Not a big fan of "read the docs to understand the dependencies."  Also feel this is an unnecessary burden on developers and is unnecessary.  I don't feel that we totally understand how developers wish the use the software and as such we're building a hyperflexible system that could theoretically do anything and in doing so makes it hard to do anything.  

As an alternative, we should be able to say "there are no dependencies! You can extend the platform via x, y, and z.  If you want to use some common services, you can load them by doing ___.  If you want to use other services, you can load them as you would any other javascript!".